src/VisualStudio/Core/Def/ProjectSystem/SdkAnalyzerAssemblyRedirector.cs

```diff
@@ -0,0 +1,190 @@
+/// </summary>
+internal class SdkAnalyzerAssemblyRedirectorCore : IAnalyzerAssemblyRedirector
+{
+    private readonly IVsActivityLog? _log;
+
+    private readonly bool _enabled;
+
+    private readonly string? _insertedAnalyzersDirectory;
+
+    /// <summary>
+    /// Map from analyzer assembly name (file name without extension) to a list of matching analyzers.
+    /// </summary>
+    private readonly ImmutableDictionary<string, List<AnalyzerInfo>> _analyzerMap;
+
+    public SdkAnalyzerAssemblyRedirectorCore(string? insertedAnalyzersDirectory, IVsActivityLog? log = null)
+    {
+        _log = log;
+        var enable = Environment.GetEnvironmentVariable("DOTNET_ANALYZER_REDIRECTING");
+        _enabled = !"0".Equals(enable, StringComparison.OrdinalIgnoreCase) && !"false".Equals(enable, StringComparison.OrdinalIgnoreCase);
+        _insertedAnalyzersDirectory = insertedAnalyzersDirectory;
+        _analyzerMap = CreateAnalyzerMap();
+    }
+
+    private ImmutableDictionary<string, List<AnalyzerInfo>> CreateAnalyzerMap()
+    {
+        if (!_enabled)
+        {
+            Log("Analyzer redirecting is disabled.");
+            return ImmutableDictionary<string, List<AnalyzerInfo>>.Empty;
+        }
+
+        var metadataFilePath = Path.Combine(_insertedAnalyzersDirectory, "metadata.json");
+        if (!File.Exists(metadataFilePath))
+        {
+            Log($"File does not exist: {metadataFilePath}");
+            return ImmutableDictionary<string, List<AnalyzerInfo>>.Empty;
+        }
+
+        var versions = JsonSerializer.Deserialize<Dictionary<string, string>>(File.ReadAllText(metadataFilePath));
+        if (versions is null || versions.Count == 0)
+        {
+            Log($"Versions are empty: {metadataFilePath}");
+            return ImmutableDictionary<string, List<AnalyzerInfo>>.Empty;
+        }
+
+        var builder = ImmutableDictionary.CreateBuilder<string, List<AnalyzerInfo>>(StringComparer.OrdinalIgnoreCase);
+
+        // Expects layout like:
+        // VsInstallDir\DotNetRuntimeAnalyzers\WindowsDesktopAnalyzers\analyzers\dotnet\System.Windows.Forms.Analyzers.dll
+        //                                     ~~~~~~~~~~~~~~~~~~~~~~~                                                     = topLevelDirectory
+        //                                                             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ = analyzerPath
+
+        foreach (var topLevelDirectory in Directory.EnumerateDirectories(_insertedAnalyzersDirectory))
+        {
+            foreach (var analyzerPath in Directory.EnumerateFiles(topLevelDirectory, "*.dll", SearchOption.AllDirectories))
+            {
+                if (!analyzerPath.StartsWith(topLevelDirectory, StringComparison.OrdinalIgnoreCase))
+                {
+                    continue;
+                }
+
+                var subsetName = Path.GetFileName(topLevelDirectory);
+                if (!versions.TryGetValue(subsetName, out var version))
```

Comment:
It'd be nice if there was a comment what the versions JSON is or looks like. There's not a file I can look at right now so I'm guessing, but I would have assumed versioning was related to the individual DLL which is odd since this path isn't depending on that yet....